fs/fs_zenfs.cc

```diff
@@ -278,10 +278,13 @@ IOStatus ZenFS::WriteEndRecord(ZenMetaLog* meta_log) {
 /* Assumes the files_mtx_ is held */
 IOStatus ZenFS::RollMetaZoneLocked() {
   std::unique_ptr<ZenMetaLog> new_meta_log, old_meta_log;
-  Zone* new_meta_zone;
+  std::shared_ptr<Zone*> pp_meta_zone(new Zone*);
```

Comment:
Why a shared pointer? The meta zones are owned by the zbd class and must live until it is destroyed. When this shared pointer is deleted, it will free the zone's memory, corrupting the meta zone list

Reply:
If we make the meta zones / io zones vectors to use shared pointers in stead of raw pointers , this could  work, but let's not do that now.

Reply:
@yhr As far as I can tell, no corruption will come of this. But I think using a `shared_ptr` here has no merit. Just use a `**Zone`. 

@MaisenbacherD `shared_ptr` only makes sense if you want to use reference count to make sure that something is destroyed when all references to that thing is dropped. Your reference counted thing here is a pointer, not an object. As far as I know, the destructor for a pointer is the same as for other primitive types, it is a no-op (except for freeing of the 8 bytes the pointer consumes). So while you do free the 8 bytes that your pointer consumes, when the `shared_ptr` is dropped, you could just allocate the pointer on the stack in the first place, instead of allocating it on heap with `new` operator.

Reply:
@metaspace _: yeah, you're right. this is a shared pointer to a pointer. no corruption, and yes, lets avoid putting things on top  of other things